packages/itwinui-react/src/core/ComboBox/ComboBoxMenu.tsx

```diff
@@ -49,18 +58,42 @@ const VirtualizedComboBoxMenu = (props: React.ComponentProps<'div'>) => {
     );
   }, [focusedIndex, menuRef]);
 
-  const { outerProps, innerProps, visibleChildren } = useVirtualization({
-    // 'Fool' VirtualScroll by passing length 1
-    // whenever there is no elements, to show empty state message
-    itemsLength: filteredOptions.length || 1,
-    itemRenderer: virtualItemRenderer,
-    scrollToIndex: focusedVisibleIndex,
-  });
+  useLayoutEffect(() => {
+    if (focusedVisibleIndex) {
+      virtualizer.scrollToIndex(focusedVisibleIndex, { align: 'start' });
+    }
+  }, [focusedVisibleIndex, virtualizer]);
```

Comment:
This useLayoutEffect feels like a hack and it doesn't work when you select an item that's in the middle of the list (below the initial view) and open the dropdown. How were we handling this  previously?

Reply:
Also the `start` behavior is not ideal. Previously we were using `nearest` (see [docs](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView#block)).

Reply:
This screen-recording shows the problem where arrow-key navigation doesn't work if the selected item is below the initial view.

https://github.com/iTwin/iTwinUI/assets/9084735/e77b9cdb-2498-4893-9010-ff9fa0c87401

Reply:
Thanks for the screen-recording. I've been able to recreate the issue on my machine and am currently looking for a fix. Definitely seems to be coming from the `scrollToIndex` function, since using the mouse to select causes the same broken state to occur when the menu is reopened. Screen-recording: 

https://github.com/iTwin/iTwinUI/assets/141063311/9eabcb70-c470-429b-a2a1-ddf108f306de

Reply:
This issue is pretty frustrating, since it seems to be coming from [here](https://github.com/TanStack/virtual/issues/216). The core issue comes from the virtualizer scrolling before the element measurement has finished, causing the scroll to go to the wrong place. There are a couple of workarounds proposed in the linked issue, using functions like `setTimeout` and `requestAnimationFrame` to delay the scroll until the measurement has happened. Both of these seem hacky to me, but there is a different solution. Adjusting the the size estimate to `36px` for the options prevents size adjustments for any options that do not include sub labels.. This fixes the issue for our specific example of a virtualized ComboBox, but wouldn't fix the problem for any ComboBoxes that use a large amount of options with sublabels.
Also started using `auto` as the alignment value, as that seems to be the closest to `nearest`. 
The keyboard navigation issues only seem to occur when the `scrollToIndex` function doesn't scroll the option into view correctly, which shouldn't happen anymore with the latest commit.
Changed the alignment of the scroll and updated the estimate size value in this [commit](https://github.com/iTwin/iTwinUI/pull/2061/commits/4e02ee85a627764fcc97a82ed650f2e581ab503f).

Reply:
Are you sure about the cause of the issue? The issue you linked was fixed in tanstack-virtual v3, so it shouldn't apply anymore.

I would think the actual cause is that you're using `useLayoutEffect`, which is never going to be called at a reliable time. You might want to instead try using a [callback ref](https://react.dev/reference/react-dom/components/common#ref-callback), which gets invoked when the DOM node is actually present. Another idea is to update some state when the measurement completes.

I also think it's okay to delay the scroll using `queueMicrotask` (for a micro-task at the end of the current task) or `setTimeout` (for next task). For this to work well, you know _what_ you're waiting for and _when_ it happens.

As for the size estimate thing, I think we could conditionally use 36 vs 48 based on some heuristics (e.g. if 3 of the first 5 items have sublabels, then use 48). But this is meant to be an "estimate", so we shouldn't rely on it.

Reply:
Tried both adding a callback ref and updating state with the `measureElement` prop passed into `useVirtualizer`, neither worked. The React components for both the outer portion of the `VirtualizedComboBoxMenu1 component and the list box that renders it both rerender whenever they are scrolled, which causes the callback ref to be constantly called whenever scrolling happens, not allowing users to freely scroll once an item has been selected. Video:

https://github.com/iTwin/iTwinUI/assets/141063311/f4edda30-f1a9-442e-8fc6-d830ba7ae86e

Updating the state does not work because updating the state with every completed measure causes too many rerenders to happen, and limiting the state update to just when the focused item is measured doesn't work because elements are only measured when they are rendered, and closing and reopening the box automatically returns to the top, causing elements far enough down the list to not be scrolled to.

Wrapping the code in the `useLayoutEffect` in a `setTimeout` does seem to work aligning items correctly on scroll regardless of the estimate size. Video: 

https://github.com/iTwin/iTwinUI/assets/141063311/3bdf6b9c-91e2-470c-9979-e336ff4871bd

Added the wrapping `setTimeout` and also added a heuristic for the estimate size in this [commit](https://github.com/iTwin/iTwinUI/pull/2061/commits/00296e8ad3db213124cd5d26b78e061b31721d1e).

Reply:
There's still an issue. If you scroll away from the current item using your mouse, arrow keys stop working.

https://github.com/iTwin/iTwinUI/assets/9084735/febb04c5-5de6-4c01-a025-d63d7225697a

Reply:
This issue was being caused by the virtualization, since the keyboard navigation code relied on being able to find the element that the `focusedIndex` pointed to, which would not be the case if users scroll away from focused elements. Fixed this by having a check for if the `ComboBox` is virtualized and if the `focusedIndex` element is not rendered, then select either the first or last rendered option, if the pressed down or up, respectively. [commit](https://github.com/iTwin/iTwinUI/pull/2061/commits/5cb826a6538b5512ef8de73ad9ec8d260a9a8c0e). Screen recording of new behavior:
https://github.com/iTwin/iTwinUI/assets/141063311/65de104d-7aa6-4c08-ab98-2b24bb98b55b

Reply:
I don't think that behavior is right. When pressing <kbd>↑</kbd> or <kbd>↓</kbd>, it should do the same thing it does for non-virtualized comboboxes, i.e. scroll back to the focused item and move focus to the corresponding adjacent item.

If this is too difficult to fix, then we can list this out as a "known issue" somewhere. After all, it is somewhat of an edge case, requiring the use of mouse and keyboard together.

Reply:
Okay. To do that, the `ComboBoxInput` component would need to be able to call the `virtualizer.scrollToIndex` function, is currently being instantiated  in `ComboBoxMenu`. I could try to instantiate the `virualizer` in the `ComboBox` component and then pass it down to `ComboBoxInput` and `ComboBoxMenu` through context or props, but that would be a pretty large refactor. Because of that, I think it might be better to create an issue for it and come back to it later. Do you agree? Also, removed the previous changes to keyboard navigation behavior in this [commit](https://github.com/iTwin/iTwinUI/pull/2061/commits/6387a29580c44968c78457adf9b9832453f44bfb).

Reply:
I agree it could be too big of a change and should be done in a separate PR. Not sure it needs an open issue though. I've created a [draft issue](https://github.com/orgs/iTwin/projects/52/views/1?sliceBy%5Bvalue%5D=iTwin%2FiTwinUI&pane=issue&itemId=68716376) where you can list out this problem. We can convert it to a public issue in the future if necessary.

Reply:
Okay, added a short description of the problem to the draft issue.